cparser/Elab.ml

```diff
@@ -983,13 +1002,18 @@ let init_int_array_wstring opt_size s =
 let check_init_type loc env a ty =
   if wrap2 valid_assignment loc env a ty then ()
   else if wrap2 valid_cast loc env a.etyp ty then
-    warning loc
-      "initializer has type@ %a@ instead of the expected type @ %a"
-      Cprint.typ a.etyp Cprint.typ ty
+    if wrap2 int_pointer_conversion loc env a.etyp ty then
+      warning loc Int_conversion
+        "incompatible integer-pointer conversion initializer has type %a instead of the expected type %a"
+         (print_typ env) a.etyp (print_typ env) ty
+    else
```

Comment:
I would be OK with 
```
   else if wrap2 int_pointer_conversion loc env ty a.etyp then
       (* specific warning *)
   else
      (* generic warning *)
```
if you think it is useful.  But the current code is OK with me too.